middle_end/flambda2/simplify/simplify_let_cont_expr.ml

```diff
@@ -919,63 +1074,99 @@ let sort_handlers data handlers =
+       parameters. This could not be done earlier, as we might not have seen
+       every use of the continuations (particularly in the recursive case) (done
+       in the [create_handler_to_rebuild] function).
+
+       - We need to compute the extra params and args related to lifting of
+       continuations. For the same reasons as unboxing of params, we need to
+       wait until here to ensure that we have seen all uses of the continuation
+       (done in the [create_handler_to_rebuild] function).
+
+       - Now that we were able to compute the extra params and args, we add this
+       information to the flow analysis, so that it can correctly compute all
+       the information we need when going up.
+
+       - We perform a strongly-connected components analysis of the
+       continuations to be able to turn mutually-recursive continuations into
+       several independant blocks of recursive or non-recursive continuations.
+       In case one of those is non-recursive, we can check whether the
+       continuation is inlinable if it is used a single time. *)
+    let cont_lifting_params = DA.cont_lifting_params dacc in
+    let handlers =
+      Continuation.Map.mapi
+        (create_handler_to_rebuild cont_lifting_params data)
+        data.handlers
+    in
+    let dacc =
+      DA.map_flow_acc dacc ~f:(fun flow_acc ->
+          Continuation.Map.fold
+            (fun cont handler flow_acc ->
+              Flow.Acc.add_extra_params_and_args cont
+                (EPA.concat ~inner:handler.invariant_extra_params_and_args
+                   ~outer:handler.extra_params_and_args)
+                flow_acc)
+            handlers flow_acc)
+    in
+    let handlers_from_the_outside_to_the_inside = sort_handlers data handlers in
+    let data : prepare_to_rebuild_handlers_data =
+      Rebuild
+        { rebuild_body = data.rebuild_body;
+          handlers_from_the_outside_to_the_inside;
+          at_unit_toplevel = data.at_unit_toplevel;
+          original_invariant_params = data.invariant_params;
+          invariant_extra_params =
+            EPA.extra_params data.invariant_extra_params_and_args
+        }
+    in
+    down_to_up dacc ~rebuild:(prepare_to_rebuild_handlers data)
+
+and prepare_dacc_for_handlers dacc ~env_at_fork ~params ~is_recursive
     ~consts_lifted_during_body continuation_sort is_exn_handler_cont uses
-    ~arg_types_by_use_id =
+    ~lifted_params ~arg_types_by_use_id =
   (* In the recursive case, [params] are actually the invariant params, as we
      prepare a common dacc for all handlers. *)
   let join_result =
+    (* We add to the continuation uses (and therefore to the handler env) the
+       params added by the eventual lifting of continuations. However, we do not
+       want those lifted params to be unboxed: it should not be beneficial
+       (since the lifted params should already include unbxoed versions). *)
+    let lifted_params, uses =
```

Comment:
This might not be true for variables bound before the continuation, but that are not argument an argument of the parent continuation.